addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/discovery/DenonMarantzDiscoveryParticipant.java

```diff
@@ -0,0 +1,163 @@
+
+    private String vendor;
+
+    private String model;
+
+    private String serial;
+
+    private String friendlyName;
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
+        return Collections.singleton(THING_TYPE_AVR);
+    }
+
+    @Override
+    public String getServiceType() {
+        return RAOP_SERVICE_TYPE;
+    }
+
+    @Override
+    public DiscoveryResult createResult(ServiceInfo serviceInfo) {
+        logger.debug("AVR found: {}", serviceInfo.getQualifiedName());
+        ThingUID uid = getThingUID(serviceInfo);
+        if (uid != null) {
+            Map<String, Object> properties = new HashMap<>(2);
+
+            if (serviceInfo.getHostAddresses().length == 0) {
+                logger.debug("Could not determine IP address for the Denon/Marantz AVR");
+                return null;
+            }
+            String host = serviceInfo.getHostAddresses()[0];
+
+            logger.debug("IP Address: {}", host);
+
+            // try a HTTP request to autoconfigure the HTTP / Telnet parameter
+            HttpClient httpClient = new HttpClient();
+            ContentResponse response;
+            boolean telnetEnable = true;
+            try {
+                httpClient.start();
+                response = httpClient.newRequest("http://" + host + "/goform/Deviceinfo.xml")
+                        .timeout(3, TimeUnit.SECONDS).send();
+                int status = response.getStatus();
+                if (status == 200) {
+                    telnetEnable = false;
+                    logger.debug("We can access the HTTP API, disabling the Telnet mode by default");
+                }
+            } catch (Exception e) {
+                logger.debug("Error when trying to access AVR using HTTP: {}, reverting to Telnet mode.",
+                        e.getMessage());
+            }
+            httpClient.destroy();
+            properties.put(DenonMarantzBindingConstants.PARAMETER_HOST, host);
+            properties.put(DenonMarantzBindingConstants.PARAMETER_TELNET_ENABLED, telnetEnable);
+            properties.put(Thing.PROPERTY_SERIAL_NUMBER, this.serial);
+            properties.put(Thing.PROPERTY_VENDOR, this.vendor);
+            properties.put(Thing.PROPERTY_MODEL_ID, this.model);
+
+            String label = this.friendlyName + " (" + this.vendor + ' ' + this.model + ")";
+            DiscoveryResult result = DiscoveryResultBuilder.create(uid).withProperties(properties).withLabel(label)
+                    .withRepresentationProperty(Thing.PROPERTY_SERIAL_NUMBER).build();
+            return result;
+
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public ThingUID getThingUID(ServiceInfo service) {
+        Matcher matcher = DENON_MARANTZ_PATTERN.matcher(service.getQualifiedName());
+        if (matcher.matches()) {
+            logger.debug("This seems like a supported Denon/Marantz AVR!");
+            this.serial = matcher.group(1).toLowerCase();
```

Comment:
What if you detect more than one? By storing the serial inside a class variable you will overwrite it in case there are 2 or more results.

Reply:
Ahh, didn't realize that! Thought the Participant would get instantiated for reach result, which obviously isn't the case. Good catch, this could have resulted in a hard to debug issue.